File: partiql-ast/src/main/java/org/partiql/ast/v1/Target.java
package org.partiql.ast.v1;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

/**
 * TODO docs, equals, hashcode
 */
public abstract class Target extends AstNode {
    @Override
    public <R, C> R accept(@NotNull AstVisitor<R, C> visitor, C ctx) {
        if (this instanceof Domain) {
            return visitor.visitTargetDomain((Domain) this, ctx);
        } else {
            throw new IllegalStateException("Unexpected value: " + this);
        }
    }

    /**
     * TODO docs, equals, hashcode
     */
    public static class Domain extends Target {
        @NotNull
        public Statement statement;

        @Nullable
        public String type;

        @Nullable
        public String format;

        public Domain(@NotNull Statement statement, @Nullable String type, @Nullable String format) {
            this.statement = statement;
            this.type = type;
            this.format = format;
        }

        @NotNull
        @Override
        public Collection<AstNode> children() {
            List<AstNode> kids = new ArrayList<>();
            kids.add(statement);
            return kids;
        }

        @Override
        public <R, C > R accept(@NotNull AstVisitor < R, C > visitor, C ctx) {
            return visitor.visitTargetDomain(this, ctx);
        }
    }
}

File: partiql-ast/src/main/kotlin/org/partiql/ast/v1/Target.kt
package org.partiql.ast.v1

/**
 * TODO docs, equals, hashcode
 */
public abstract class Target : AstNode() {
    public override fun <R, C> accept(visitor: AstVisitor<R, C>, ctx: C): R = when (this) {
        is Domain -> visitor.visitTargetDomain(this, ctx)
        else -> throw NotImplementedError()
    }

    /**
     * TODO docs, equals, hashcode
     */
    public class Domain(
        @JvmField
        public var statement: Statement,
        @JvmField
        public var type: String?,
        @JvmField
        public var format: String?,
    ) : Target() {
        public override fun children(): Collection<AstNode> {
            val kids = mutableListOf<AstNode?>()
            kids.add(statement)
            return kids.filterNotNull()
        }

        public override fun <R, C> accept(visitor: AstVisitor<R, C>, ctx: C): R =
            visitor.visitTargetDomain(this, ctx)
    }
}
